staging/src/k8s.io/apiserver/pkg/cel/lazy/lazy.go

```diff
@@ -0,0 +1,180 @@
+	v, found := m.Find(key)
+	if v != nil && types.IsUnknownOrError(v) {
+		return v
+	}
+	return types.Bool(found)
+}
+
+// Iterator returns an iterator that does nothing
+func (m *MapValue) Iterator() traits.Iterator {
+	return &dummyIterator{}
+}
+
+// Size returns the number of currently known fields
+func (m *MapValue) Size() ref.Val {
+	return types.Int(len(m.callbacks))
+}
+
+// ConvertToNative returns an error because it is disallowed
+func (m *MapValue) ConvertToNative(typeDesc reflect.Type) (any, error) {
+	return nil, fmt.Errorf("disallowed conversion from %q to %q", m.typeValue.TypeName(), typeDesc.Name())
+}
+
+// ConvertToType converts the map to the given type.
+// Only its own type and "Type" type are allowed.
+func (m *MapValue) ConvertToType(typeVal ref.Type) ref.Val {
+	switch typeVal {
+	case m.typeValue:
+		return m
+	case types.TypeType:
+		return m.typeValue
+	}
+	return types.NewErr("disallowed conversion from %q to %q", m.typeValue.TypeName(), typeVal.TypeName())
+}
+
+// Equal returns true if the other object is the same pointer-wise.
+func (m *MapValue) Equal(other ref.Val) ref.Val {
+	otherMap, ok := other.(*MapValue)
+	if !ok {
+		return types.MaybeNoSuchOverloadErr(other)
+	}
+	return types.Bool(m == otherMap)
+}
+
+// Type returns its registered type.
+func (m *MapValue) Type() ref.Type {
+	return m.typeValue
+}
+
+// Value is not allowed.
+func (m *MapValue) Value() any {
+	return types.NoSuchOverloadErr()
+}
+
+// resolveField resolves the field. Calls the callback if the value is not yet stored.
+func (m *MapValue) resolveField(name string) ref.Val {
+	v, seen := m.values[name]
+	if seen {
+		return v
+	}
+	f := m.callbacks[name]
+	v = f(m)
+	m.values[name] = v
+	return v
+}
+
+func (m *MapValue) Find(key ref.Val) (ref.Val, bool) {
+	n, ok := key.(types.String)
+	if !ok {
+		return types.MaybeNoSuchOverloadErr(n), true
+	}
+	name, ok := cel.Unescape(n.Value().(string))
+	if !ok {
+		return nil, false
+	}
+	if _, exists := m.callbacks[name]; !exists {
+		return nil, false
+	}
+	return m.resolveField(name), true
+}
+
+func (m *MapValue) Get(key ref.Val) ref.Val {
+	v, found := m.Find(key)
+	if found {
+		return v
+	}
+	return types.ValOrErr(key, "no such key: %v", key)
+}
+
+type dummyIterator struct {
+}
+
+func (i *dummyIterator) ConvertToNative(typeDesc reflect.Type) (any, error) {
+	return nil, fmt.Errorf("disallowed conversion to %q", typeDesc.Name())
+}
+
+func (i *dummyIterator) ConvertToType(typeValue ref.Type) ref.Val {
+	return types.NewErr("disallowed conversion o %q", typeValue.TypeName())
+}
+
+func (i *dummyIterator) Equal(other ref.Val) ref.Val {
+	return types.Bool(false)
+}
+
+func (i *dummyIterator) Type() ref.Type {
+	return types.NullType
+}
+
+func (i *dummyIterator) Value() any {
+	return struct{}{}
+}
+
+func (i *dummyIterator) HasNext() ref.Val {
+	return types.Bool(false)
+}
+
+func (i *dummyIterator) Next() ref.Val {
+	return types.NewErr("iterating a lazy map is disallowed.")
```

Comment:
It might be useful when debugging, e.g. `messageExpression: 'variables.map(k, v, k + "=" + string(v)).join(", ")'`.  I realize iterating would be an expensive operation, but maybe we should support it for consistency? Is it tricky to implement?

Reply:
It is not hard to implement as long as the lazy map is append-only. Let me add just for the convenience.

Reply:
Wait a sec. `.map` cannot be used on a map. 

https://github.com/google/cel-spec/blob/master/doc/langdef.md#macros

e.map(x, t) transforms a list e by taking each element x to the function given by the expression t, which can use the variable x. For instance, [1, 2, 3].map(n, n * n) evaluates to [1, 4, 9]. Any evaluation error for any element causes the macro to raise an error. **The map() macro is not supported when e is a map.**

Is there an example for iterating a map so I can use it in the tests?

Reply:
`.all` should do it.